vendor/bat-native-ledger/src/bat/ledger/internal/database/database_unblinded_token.cc

```diff
@@ -504,6 +547,110 @@ void DatabaseUnblindedToken::MarkRecordListAsSpent(
   ledger_->RunDBTransaction(std::move(transaction), transaction_callback);
 }
 
+void DatabaseUnblindedToken::MarkRecordListAsReserved(
+    const std::vector<std::string>& ids,
+    const std::string& redeem_id,
+    ledger::ResultCallback callback) {
+  if (ids.empty()) {
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+
+  auto transaction = ledger::DBTransaction::New();
+
+  const std::string id_values = GenerateStringInCase(ids);
+
+  std::string query = base::StringPrintf(
+      "UPDATE %s SET redeem_id = ?, reserved_at = ? "
+      "WHERE ( "
+        "SELECT COUNT(*) FROM %s "
+        "WHERE reserved_at = 0 AND token_id IN (%s) "
+      ") = ? AND token_id IN (%s)",
+      kTableName,
+      kTableName,
+      id_values.c_str(),
+      id_values.c_str());
+
+  auto command = ledger::DBCommand::New();
+  command->type = ledger::DBCommand::Type::RUN;
+  command->command = query;
+
+  BindString(command.get(), 0, redeem_id);
+  BindInt64(command.get(), 1, braveledger_time_util::GetCurrentTimeStamp());
+  BindInt64(command.get(), 2, ids.size());
```

Comment:
could you please explain why do we need this one?

Reply:
I want to ensure that the query only succeeds if exactly the requested number of rows are affected. Otherwise it seems like we could ask to reserve 3 tokens (for example) and this update will still succeed if it only reserves 2 tokens. Am I thinking about it wrong? It doesn't seem like you can just check this value in the callback, as by then it looks like the transaction has ended.

Reply:
sounds good, nice idea :)